src/bulk_insert/bulk_insert.c

```diff
@@ -7,340 +7,186 @@
+    LabelStore_UpdateSchema(store, *prop_count, prop_keys);
+
+    return prop_keys;
 }
 
-// Parse label from argv.
-RedisModuleString** _BulkInsert_Parse_Label(RedisModuleCtx *ctx, LabelDesc *label,
-                                             RedisModuleString **argv, int *argc) {
-    // Minimum of 3 arguments: label name, number of labeled nodes and attribute count.
-    if(*argc < 3) {
-        _BulkInsert_Reply_With_Syntax_Error(ctx, "Bulk insert format error, missing label parameters.");
-        return NULL;
-    }
-    *argc -= 3;
-
-    // Label's name.
-    label->label = RedisModule_StringPtrLen(*argv++, NULL);
-
-    // Number of nodes under this label.
-    long long node_count = 0;
-    if (RedisModule_StringToLongLong(*argv++, &node_count) != REDISMODULE_OK) {
-        _BulkInsert_Reply_With_Syntax_Error(ctx, "Bulk insert format error, failed to parse labeled node count.");
-        return NULL;
-    }
+// Read an SIValue from the data stream and update the index appropriately
+static inline SIValue _BulkInsert_ReadProperty(const char *data, size_t *data_idx) {
+    SIValue v;
+    TYPE t = data[*data_idx];
+    *data_idx += 1;
+    if (t == BI_NULL) {
+        // TODO This property will currently get entered with a NULL key.
+        // Update so that the entire key-value pair is omitted from the node
+        v = SI_NullVal();
+    } else if (t == BI_BOOL) {
+        bool b = data[*data_idx];
+        *data_idx += 1;
+        v = SI_BoolVal(b);
+    } else if (t == BI_NUMERIC) {
+        double d = *(double*)&data[*data_idx];
+        *data_idx += sizeof(double);
+        v = SI_DoubleVal(d);
+    } else if (t == BI_STRING) {
+        char *s = rm_strdup(data + *data_idx);
+        *data_idx += strlen(s) + 1;
+        // Ownership of the string will be passed to the GraphEntity properties
+        v = SI_TransferStringVal(s);
+    } else {
+        assert(0);
+    }
+    return v;
+}
 
-    label->node_count = node_count;
+int _BulkInsert_ProcessNodeFile(RedisModuleCtx *ctx, GraphContext *gc, const char *data, size_t data_len) {
+    size_t data_idx = 0;
 
-    // Label's attribute count.
-    long long attribute_count = 0;
-    if(RedisModule_StringToLongLong(*argv++, &attribute_count) != REDISMODULE_OK) {
-        _BulkInsert_Reply_With_Syntax_Error(ctx, "Bulk insert format error, failed to parse label attributes count.");
-        return NULL;
-    }
-    label->attribute_count = attribute_count;
+    int label_id;
+    unsigned int prop_count;
+    char **prop_keys = _BulkInsert_ReadHeader(gc, STORE_NODE, data, &data_idx, &label_id, &prop_count);
 
-    if(attribute_count > 0) {
-        if(*argc < attribute_count) {
-            _BulkInsert_Reply_With_Syntax_Error(ctx, "Bulk insert format error, missing label attributes.");
-            return NULL;
-        }
-        *argc -= attribute_count;
+    // Reusable buffer for storing properties of each node
+    SIValue values[prop_count];
```

Comment:
I don't know what kind of fast handling we can introduce, as looping over rows is necessary no matter what.
prop_count can't be 0 for nodes, though it can be for relations. This line may actually be unsafe, since it's permissible to `malloc(0)`, but I think a stack allocation must have a size according to the standard. I don't know if any compilers have an issue with that, though. Thoughts?

Reply:
disregard my comment